linker tests are currently broken on linux and osx

Running the linker tests on non-Windows platforms currently fails with:

```
/home/eerhardt/git/runtime/.dotnet/sdk/5.0.100-rc.2.20479.15/Microsoft.Common.CurrentVersion.targets(4814,5): error MSB3030: Could not copy the file "/home/eerhardt/git/runtime/artifacts/bin/trimmingTests/projects/Microsoft.Extensions.DependencyInjection.TrimmingTests/ActivatorUtilitiesTests/linux-x64/obj/Release/net6.0/linux-x64/apphost" because it was not found. [/home/eerhardt/git/runtime/artifacts/bin/trimmingTests/projects/Microsoft.Extensions.DependencyInjection.TrimmingTests/ActivatorUtilitiesTests/linux-x64/project.csproj]
```

The linker test project cannot be built because it can't find the `AppHostIntermediatePath` file. The file doesn't exist because it isn't being created by `_CreateAppHost`, since the `$(AppHostSourcePath)` property is empty.

Here is a binlog of the failure:

[msbuild.binlog.zip](https://github.com/dotnet/runtime/files/5527116/msbuild.binlog.zip)

While investigating this break, it appears we have some incorrect code added into the linker targets files:

https://github.com/dotnet/runtime/blob/2db717954a3fd9ca5119143ccc633fef3dfa3bb8/eng/testing/linker/SupportFiles/Directory.Build.targets#L36-L53

This is incorrect because:

1. We need to be using the framework that we just built (i.e. what is in `artifacts`). These tests need to run on the current code changes, not what is in some pre-compiled version.
2. We shouldn't be hard-coding these versions into a random targets file. They need to be centralized, and kept up to date, if we do need them.

We should fix the linker-tests so they can be re-enabled in CI.

cc @joperezr @Anipik @ericstj

--- a/Directory.Build.targets
+++ b/Directory.Build.targets
@@ -29,19 +29,19 @@
   -->
   <ItemGroup>
     <KnownFrameworkReference Include="Microsoft.NETCore.App"
-      DefaultRuntimeFrameworkVersion="5.0.0-rc.1.20451.14"
+      DefaultRuntimeFrameworkVersion="$(MicrosoftNETCoreAppVersion)"
       IsTrimmable="true"
-      LatestRuntimeFrameworkVersion="5.0.0-rc.1.20451.14"
+      LatestRuntimeFrameworkVersion="$(MicrosoftNETCoreAppVersion)"
       RuntimeFrameworkName="Microsoft.NETCore.App"
       RuntimePackNamePatterns="Microsoft.NETCore.App.Runtime.**RID**"
       RuntimePackRuntimeIdentifiers="linux-arm;linux-arm64;linux-musl-arm64;linux-musl-x64;linux-x64;osx-x64;rhel.6-x64;tizen.4.0.0-armel;tizen.5.0.0-armel;win-arm;win-arm64;win-x64;win-x86;ios-arm64;ios-arm;ios-x64;ios-x86;tvos-arm64;tvos-x64;android-arm64;android-arm;android-x64;android-x86;browser-wasm"
       TargetFramework="$(NetCoreAppCurrent)"
       TargetingPackName="Microsoft.NETCore.App.Ref"
-      TargetingPackVersion="5.0.0-rc.1.20451.14" />
+      TargetingPackVersion="$(MicrosoftNETCoreAppVersion)" />
 
     <KnownAppHostPack Include="Microsoft.NETCore.App" 
       AppHostPackNamePattern="Microsoft.NETCore.App.Host.**RID**"
-      AppHostPackVersion="5.0.0-rc.1.20451.14"
+      AppHostPackVersion="$(MicrosoftNETCoreAppVersion)"
       AppHostRuntimeIdentifiers="linux-arm;linux-arm64;linux-musl-arm64;linux-musl-x64;linux-x64;osx-x64;rhel.6-x64;tizen.4.0.0-armel;tizen.5.0.0-armel;win-arm;win-arm64;win-x64;win-x86"
       TargetFramework="$(NetCoreAppCurrent)" />
   </ItemGroup>

--- a/eng/testing/linker/SupportFiles/Directory.Build.props
+++ b/eng/testing/linker/SupportFiles/Directory.Build.props
@@ -8,4 +8,8 @@
     <PlatformManifestFile />
     <DisableImplicitFrameworkReferences>true</DisableImplicitFrameworkReferences>
   </PropertyGroup>
+
+  <ItemGroup>
+    <FrameworkReference Include="Microsoft.NETCore.App" />
+  </ItemGroup>
 </Project>

--- a/eng/testing/linker/SupportFiles/Directory.Build.targets
+++ b/eng/testing/linker/SupportFiles/Directory.Build.targets
@@ -1,18 +1,20 @@
 <Project>
   <Target Name="RestoreProject">
     <MSBuild Projects="$(MSBuildProjectFullPath)"
-             Properties="Configuration=$(Configuration)"
+             Properties="Configuration=$(Configuration);ForceEvaluation=true"
              Targets="Restore" />
   </Target>
 
-  <Target Name="AddFrameworkReference">
+  <Target Name="RemoveRuntimePackFromDownloadItem"
+          AfterTargets="ProcessFrameworkReferences">
     <ItemGroup>
-      <FrameworkReference Include="Microsoft.NETCore.App" />
+      <PackageDownload Remove="@(PackageDownload)"
+                       Condition="$([System.String]::Copy('%(Identity)').StartsWith('Microsoft.NETCore.App.Runtime'))" />
     </ItemGroup>
   </Target>
 
   <Target Name="UpdateRuntimePack"
-          DependsOnTargets="AddFrameworkReference;ResolveFrameworkReferences">
+          DependsOnTargets="ResolveFrameworkReferences">
     <ItemGroup>
       <ResolvedRuntimePack Update="@(ResolvedRuntimePack)" PackageDirectory="$(RuntimePackDir)" />
       <ResolvedTargetingPack Update="@(ResolvedTargetingPack)" Path="$(TargetingPackDir)" />
@@ -31,23 +33,23 @@
     </ItemGroup>
   </Target>
 
-  <Target Name="PublishTrimmed" DependsOnTargets="RestoreProject;UpdateRuntimePack;Publish" />
+  <Target Name="PublishTrimmed" DependsOnTargets="ProcessFrameworkReferences;RestoreProject;UpdateRuntimePack;Publish" />
 
   <ItemGroup>
     <KnownFrameworkReference Include="Microsoft.NETCore.App"
-      DefaultRuntimeFrameworkVersion="5.0.0-rc.1.20451.14"
+      DefaultRuntimeFrameworkVersion="$(MicrosoftNETCoreAppVersion)"
       IsTrimmable="true"
-      LatestRuntimeFrameworkVersion="5.0.0-rc.1.20451.14"
+      LatestRuntimeFrameworkVersion="$(MicrosoftNETCoreAppVersion)"
       RuntimeFrameworkName="Microsoft.NETCore.App"
       RuntimePackNamePatterns="Microsoft.NETCore.App.Runtime.**RID**"
       RuntimePackRuntimeIdentifiers="linux-arm;linux-arm64;linux-musl-arm64;linux-musl-x64;linux-x64;osx-x64;rhel.6-x64;tizen.4.0.0-armel;tizen.5.0.0-armel;win-arm;win-arm64;win-x64;win-x86;ios-arm64;ios-arm;ios-x64;ios-x86;tvos-arm64;tvos-x64;android-arm64;android-arm;android-x64;android-x86;browser-wasm"
       TargetFramework="net6.0"
       TargetingPackName="Microsoft.NETCore.App.Ref"
-      TargetingPackVersion="5.0.0-rc.1.20451.14" />
+      TargetingPackVersion="$(MicrosoftNETCoreAppVersion)" />
 
     <KnownAppHostPack Include="Microsoft.NETCore.App" 
       AppHostPackNamePattern="Microsoft.NETCore.App.Host.**RID**"
-      AppHostPackVersion="5.0.0-rc.1.20451.14"
+      AppHostPackVersion="$(MicrosoftNETCoreAppVersion)"
       AppHostRuntimeIdentifiers="linux-arm;linux-arm64;linux-musl-arm64;linux-musl-x64;linux-x64;osx-x64;rhel.6-x64;tizen.4.0.0-armel;tizen.5.0.0-armel;win-arm;win-arm64;win-x64;win-x86"
       TargetFramework="net6.0" />
   </ItemGroup>

--- a/eng/testing/linker/project.csproj.template
+++ b/eng/testing/linker/project.csproj.template
@@ -7,6 +7,7 @@
     <RuntimePackDir>{RuntimePackDir}</RuntimePackDir>
     <TargetingPackDir>{TargetingPackDir}</TargetingPackDir>
     <NETCoreAppMaximumVersion>{NetCoreAppMaximumVersion}</NETCoreAppMaximumVersion>
+    <MicrosoftNETCoreAppVersion>{MicrosoftNETCoreAppVersion}</MicrosoftNETCoreAppVersion>
 
     <!-- These can be removed once we get an SDK with https://github.com/mono/linker/pull/1385. -->
     <LinkerNoWarn>IL2026</LinkerNoWarn>

--- a/eng/testing/linker/trimmingTests.targets
+++ b/eng/testing/linker/trimmingTests.targets
@@ -2,7 +2,6 @@
   <PropertyGroup>
     <TrimmingTestDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsBinDir)', 'trimmingTests'))</TrimmingTestDir>
     <TrimmingTestProjectsDir>$([MSBuild]::NormalizeDirectory('$(TrimmingTestDir)', 'projects'))</TrimmingTestProjectsDir>
-    <TestDotNetPath>$([MSBuild]::NormalizePath('$(DotNetRoot)', 'dotnet'))</TestDotNetPath>
     <ProjectTemplate>$(MSBuildThisFileDirectory)project.csproj.template</ProjectTemplate>
   </PropertyGroup>
 
@@ -73,6 +72,7 @@
     <MakeDir Directories="$(_projectDir)" />
     <WriteLinesToFile File="$(_projectFile)"
                       Lines="$([System.IO.File]::ReadAllText('$(ProjectTemplate)')
+                                                 .Replace('{MicrosoftNETCoreAppVersion}', '$(MicrosoftNETCoreAppVersion)')
                                                  .Replace('{NetCoreAppCurrent}', '$(NetCoreAppCurrent)')
                                                  .Replace('{NetCoreAppMaximumVersion}', '$(NetCoreAppMaximumVersion)')
                                                  .Replace('{RuntimePackDir}', '%(TestConsoleApps.RuntimePackDirectory)')
@@ -93,7 +93,7 @@
   <Target Name="PublishTrimmedProjects"
           DependsOnTargets="GenerateProjects">
     <PropertyGroup>
-      <TestPublishTrimmedCommand>"$(TestDotNetPath)"</TestPublishTrimmedCommand>
+      <TestPublishTrimmedCommand>"$(DotNetTool)"</TestPublishTrimmedCommand>
       <TestPublishTrimmedCommand>$(TestPublishTrimmedCommand) build /t:PublishTrimmed</TestPublishTrimmedCommand>
       <TestPublishTrimmedCommand>$(TestPublishTrimmedCommand) /nr:false</TestPublishTrimmedCommand>
       <TestPublishTrimmedCommand>$(TestPublishTrimmedCommand) /warnaserror</TestPublishTrimmedCommand>

